GH-98831: "Generate" the interpreter

gvanrossum opened the pull request:
### Overview

This is the first stage of a project to generate the interpreter, in particular the cases of the main switch. The references below explain the why. This PR introduces the following:

- Two new files in the Python subdirectory, [`bytecodes.c`](https://github.com/gvanrossum/cpython/blob/generate-ceval-switch/Python/bytecodes.c) and [`generated_cases.c.h`](https://github.com/gvanrossum/cpython/blob/generate-ceval-switch/Python/generated_cases.c.h). (**UPDATE:** cases.h -> generated_cases.c.h.)
  - **bytecodes.c** contains [instruction definitions](https://github.com/faster-cpython/ideas/blob/main/3.12/interpreter_definition.md). These almost look like the old switch cases, but instead of `TARGET(INAME) { ... }` they have `inst(INAME, stack_effect) { ... }`, and the `DISPATCH()` macro call at the end of the block can be omitted. (The stack effect is not yet used by the generator, and it is not always correct.) At the top and bottom of bytecodes.c is some [dummy C code](https://github.com/faster-cpython/ideas/issues/477) that makes look like valid C code to editors like VS Code.
  - **generated_cases.c.h** is a generated file containing the switch cases. It is currently identical to the switch cases in `ceval.c`.

- A new bunch of tooling, in [`Tools/cases_generator`](https://github.com/gvanrossum/cpython/tree/generate-ceval-switch/Tools/cases_generator). This includes:
  - ~A _temporary_ script,` extract_cases.py`, that extracts the cases from `ceval.c` and writes them, in the form of instruction definitions, to `bytecodes.c`.~
  - A script, `generate_cases.py`, that reads the instruction definitions from `bytecodes.c` and regenerates the cases, writing them to `generated_cases.c.h`.
  - ~A parser for a subset of C that turns blocks of C code into an AST, so we can do transformations at the AST level. (There is currently one example of this -- the function [`always_exits()`](https://github.com/gvanrossum/cpython/blob/generate-ceval-switch/Tools/cases_generator/generate_cases.py#L59) determines recursively whether a block is guaranteed to exit, using a custom definition of "exit" that includes macro calls like `DISPATCH()` and `GO_TO_INSTRUCTION()`.)~

- Changes to ceval.c so that it uses `#include "generated_cases.c.h"`. For now I've left the original switch cases in ceval.c, inside `#else` ... `#endif`, so that we can compare the original switch cases to the generated switch cases to validate the toolchain. However, I plan to rip these out either once the PR is (nearly) approved, or perhaps in a quick follow-up PR.

- A new [target](https://github.com/gvanrossum/cpython/blob/01b3735d2a6540a8ab419e6ffb1f579ac068e94a/Makefile.pre.in#L1448) in `Makefile.pre.in` to auto-generate `generated_cases.c.h` from `bytecodes.c`, using `make regen-cases`.
  - Currently missing is similar tooling on Windows. For now, you can only regenerate the cases on UNIX. They are checked into the repo, so Windows should still compile.

### References

- https://github.com/faster-cpython/ideas/issues/5
- https://github.com/faster-cpython/ideas/issues/16
- https://github.com/faster-cpython/ideas/issues/454
- https://github.com/faster-cpython/ideas/issues/477
- https://github.com/faster-cpython/ideas/issues/479
- https://github.com/faster-cpython/ideas/issues/481

### Metadata

<!-- gh-issue-number: gh-98831 -->
* Issue: gh-98831
<!-- /gh-issue-number -->

gvanrossum committed:
New script to generate the ceval switch

This doesn't work yet, just a checkpoint. I'm going to bed now.

It's going to be a long slog.

gvanrossum committed:
Add 'End regular instructions' marker

gvanrossum committed:
Extract to a format closer to interpreter_definition.md

gvanrossum committed:
Switch to generating new DSL syntax

gvanrossum committed:
Indent output; Always generate stack effect syntax; use argparse

gvanrossum committed:
Check jump flag; tweak begin and end of the script a bit

gvanrossum committed:
Fix incorrect output for ??--??

gvanrossum committed:
Script to generate cases.h from bytecodes.inst

gvanrossum committed:
Write bytecodes.inst into Python directory by default

gvanrossum committed:
Initial bytecodes.inst

gvanrossum committed:
Initial cases.h

gvanrossum committed:
Changes to ceval.c to use cases.h

gvanrossum committed:
Changes to Makefile.pre.in to generate cases.h

gvanrossum committed:
Move #endif around

gvanrossum committed:
Write families to .inst file

(Don't do anything with them yet.)

gvanrossum committed:
Derive stack effect for super-instructions and specializations

gvanrossum committed:
Regenerated bytecodes.inst with fewer errors

gvanrossum committed:
Support array stack effects

gvanrossum committed:
Array stack effects are now in bytecodes.inst

gvanrossum committed:
Add Mark's C lexer

gvanrossum committed:
Start writing a real parser

(This is a checkpoint, doesn't work yet.)

gvanrossum committed:
Move cases infrastructure to its own directory

gvanrossum committed:
Pristine copy of lexer from Mark's gist

https://gist.github.com/markshannon/db7ab649440b5af765451bb77c7dba34#gistcomment-4334862

gvanrossum committed:
Fix tiny bits in lexer.py

gvanrossum committed:
Move c_lexer.c to lexer.c

gvanrossum committed:
Complete parser (not using it yet):

gvanrossum committed:
Complete the parser and use it

gvanrossum committed:
Use a dataclass for Token; move stuff around to fix tokens

gvanrossum committed:
Make a nice syntax error in lexer.py

gvanrossum committed:
More handy stuff for raising syntax errors

gvanrossum committed:
Fix family definitions per spec

gvanrossum committed:
Rename bytecodes.{inst,c} in prep for better tooling support

gvanrossum committed:
Use a template to generate top/bottom of bytecodes.c

gvanrossum committed:
Add more to bytecodes_template.c

gvanrossum committed:
Let generate_cases.py add the DISPATCH() calls

There are a few redundant DISPATCH() calls added,
but those seem harmless (the optimizer should remove them),
and we will make things more sophisticated later.

gvanrossum committed:
Refactor the parser in preparation of actually parsing C

gvanrossum committed:
Handle PERIOD properly

gvanrossum committed:
CHECKPOINT -- the start of a better C parser

gvanrossum committed:
Merge remote-tracking branch 'origin/main' into generate-ceval-switch

gvanrossum committed:
Add some type hints to lexer.py

gvanrossum committed:
Split cparser into sparser and eparser

gvanrossum committed:
Identifiers are types too (-ish)

gvanrossum committed:
Begin better treatment of types

gvanrossum committed:
Begin better declarations

gvanrossum committed:
Support any number of postfix operators

gvanrossum committed:
Split declaration; support 'for (int i = 0;...)'

gvanrossum committed:
Use CHARACTER, not CHAR, for '.' token

gvanrossum committed:
Support string and character literals

gvanrossum committed:
Refactor decl_stmt some more

gvanrossum committed:
Switch statement (and case and default)

gvanrossum committed:
String literal concatenation

gvanrossum committed:
Fix hex and octal numbers

gvanrossum committed:
Hack conditional operator

gvanrossum committed:
More numeric types, e.g. unsigned long

gvanrossum committed:
Support array declarations and initializers

gvanrossum committed:
Fix typo in main

gvanrossum committed:
Support some function types in casts

NOTE: Types are still a mess (deservedly :-).

gvanrossum committed:
Support function pointer declarations

gvanrossum committed:
Make Context repr more compact

gvanrossum committed:
Make contextual more generic

gvanrossum committed:
Make Parser derive from SParser, and add test main

gvanrossum committed:
CHECKPOINT: generate cases the new way

(This doesn't yet insert DISPATCH() into the code.)

gvanrossum committed:
Insert DISPATCH() unless block always exits

gvanrossum committed:
More accuracy in DISPATCH() addition

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

gvanrossum committed:
Merge remote-tracking branch 'origin/main' into generate-ceval-switch

gvanrossum committed:
Update generated cases due to END_FOR refactor

gvanrossum committed:
Add some TODO comments to lexer.py

gvanrossum committed:
Add README.md to Tools/cases_generator

gvanrossum committed:
Fix trailing whitespace

gvanrossum committed:
Fix crash in require() at EOF

gvanrossum committed:
More compact Token.__repr__()

gvanrossum committed:
Proper operator priorities, add comma operator

Also make Name.__repr__() and Context.__repr__() more compact;
and various other minor cleanups.

markshannon commented on Python/bytecodes.c:
```diff
@@ -0,0 +1,3795 @@
+    PyObject **stack_pointer
+)
+{
+    switch (opcode) {
```
If we `#define` `inst(NAME, args)` as `void NAME(void)`, there is no need for the dummy function or `switch`.
That way all the instructions can be un-indented.

gvanrossum replied:
The indentation is not the issue. If each instruction lived in its own function, references to variables like `frame` or `eval_breaker` would all get red wiggly lines in VS Code. Every single variable I listed as a parameter to `dummy_func()` above is needed to avoid wiggly lines. There are also some labels at the bottom of the file that are needed as the target of `goto` labels.

markshannon commented on Python/bytecodes.c:
```diff
@@ -0,0 +1,3795 @@
+            PUSH(value);
+        }
+
+        inst(LOAD_FAST_CHECK, ( -- __0)) {
```
This is incorrect. Either the stack effect should be present *or* the `PUSH` in the code below.
Having both implies that the generated code would look like this:
```
    ...
    PUSH(value);
    PUSH(__0);
```
It also prevents manually converting from the procedural form to the declarative form.

gvanrossum replied:
I can easily get rid of the stack effect, but my thinking here is that the code generator would ignore "dummy" stack effects that begin with `__` (currently it ignores all stack effects). We could then later use these as a starting point for manual conversion to correct stack effects. It's true that technically the dummy stack effects are redundant because the real stack effect is present in the code, and in some cases the dummy stack effect is incorrect (because it's derived from the `stack_effect()` function, which doesn't distinguish between e.g. `(x -- y, z)` and `( -- z)`).

I had hoped it would help, not hinder with conversion, but if you think it's confusing I'll get rid of it.

markshannon commented on Python/bytecodes.c:
```diff
@@ -0,0 +1,3795 @@
+
+family(binary_op) = BINARY_OP + BINARY_OP_ADAPTIVE + BINARY_OP_ADD_FLOAT + BINARY_OP_ADD_INT + BINARY_OP_ADD_UNICODE + BINARY_OP_INPLACE_ADD_UNICODE + BINARY_OP_MULTIPLY_FLOAT + BINARY_OP_MULTIPLY_INT + BINARY_OP_SUBTRACT_FLOAT + BINARY_OP_SUBTRACT_INT;
+family(binary_subscr) = BINARY_SUBSCR + BINARY_SUBSCR_ADAPTIVE + BINARY_SUBSCR_DICT + BINARY_SUBSCR_GETITEM + BINARY_SUBSCR_LIST_INT + BINARY_SUBSCR_TUPLE_INT;
+family(call) = CALL + CALL_ADAPTIVE + CALL_PY_EXACT_ARGS + CALL_PY_WITH_DEFAULTS + CALL_BOUND_METHOD_EXACT_ARGS + CALL_BUILTIN_CLASS + CALL_BUILTIN_FAST_WITH_KEYWORDS + CALL_METHOD_DESCRIPTOR_FAST_WITH_KEYWORDS + CALL_NO_KW_BUILTIN_FAST + CALL_NO_KW_BUILTIN_O + CALL_NO_KW_ISINSTANCE + CALL_NO_KW_LEN + CALL_NO_KW_LIST_APPEND + CALL_NO_KW_METHOD_DESCRIPTOR_FAST + CALL_NO_KW_METHOD_DESCRIPTOR_NOARGS ...
```
Could we improve the formatting?
```C
family(binary_op) =
    BINARY_OP + BINARY_OP_ADAPTIVE + BINARY_OP_ADD_FLOAT + BINARY_OP_ADD_INT +
    BINARY_OP_ADD_UNICODE + BINARY_OP_INPLACE_ADD_UNICODE + BINARY_OP_MULTIPLY_FLOAT +
    BINARY_OP_MULTIPLY_INT + BINARY_OP_SUBTRACT_FLOAT + BINARY_OP_SUBTRACT_INT;
```

gvanrossum replied:
Sure, that's simple enough. :-)

markshannon commented on Tools/cases_generator/eparser.py:
```diff
@@ -0,0 +1,489 @@
+# C expression parser
```
This is out of scope for our current needs. All we need for now is to count braces.
I'm sure it be useful in future, but maybe stash it somewhere else for now.

gvanrossum replied:
We'll need it as soon as we're going to generate super-instructions from basic opcodes, since stitching blocks of C code together is not _entirely_ trivial.

markshannon commented on Tools/cases_generator/extract_cases.py:
```diff
@@ -0,0 +1,236 @@
+"""Extract the main interpreter switch cases."""
```
This is unnecessary. We don't need to extract the instructions from the interpreter, we need to create the interpreter from the instructions.

gvanrossum replied:
I know, I know. The README.md explains this (I hope). It's a throwaway script, but I want to throw it awa a few weeks after the PR has been merged.

markshannon commented on Tools/cases_generator/sparser.py:
```diff
@@ -0,0 +1,335 @@
+# C statement parser
```
Like the expression parser, we don't need this. Maybe in the future.

gvanrossum replied:
I needed it for a very straightforward transformation, which I implemented as an example of what the generator can do: automatically insert `DISPATCH()` macros at the end of an instruction unless the instruction already ends in something else that forces the control flow to go elsewhere. I first had a version that just skipped braces, and it was impossible to get everything right. Look for `always_exits()` in generate_cases.py.

markshannon commented on Tools/cases_generator/plexer.py:
```diff
@@ -0,0 +1,104 @@
+import lexer as lx
```
Why do we need two lexers?

gvanrossum replied:
Different abstraction levels. This file provides an OO interface on top of your tokenize() function that lets the parser backtrack easily.

markshannon commented on Python/ceval.c:
```diff
@@ -1162,6 +1162,23 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
+#include "cases.h"
+#else
+
         TARGET(NOP) {
```
Why not move all the instruction definitions to `bytecode.c` and put `#include "generated_cases.c.h."` here?

gvanrossum replied:
I agree with you (in https://github.com/python/cpython/pull/98830#issuecomment-1295732911). I'd like to do this last, just before merging the PR, so I can re-run the extraction script after merging main into the PR.

markshannon commented on Tools/cases_generator/bytecodes_template.c:
```diff
@@ -0,0 +1,43 @@
+#include "Python.h"
```
This file is only needed when extracting the instructions from ceval.c?

gvanrossum replied:
It's an implementation detail for extract_cases.py, which is itself redundant once the PR is merged (though I'd like to keep it in the repo for a few weeks to help people who have a pending PR that changes instruction cases).

markshannon reviewed (commented):
This seems too complex for our current needs. Although the various bits and pieces will almost certainly be useful in the future.

What I had hoped for was little more than copy-and-paste moving of `bytecodes.c` to some `generated.c.h` which is  `#include`d in `ceval.c`

We will want to get more sophisticated than that soon enough, which is why we want to parse the instructions file. But initially we just want the minimum thing that produces an interpreter.

1. Cut and paste the `TARGET()`s from `ceval.c` into `bytecodes.c`
2. Put `#include "generated.c.h" where the instructions were.
3. Add sufficient preamble to `bytecodes.c` so that IDEs are happy
4. Generate `generated.c.h` from  `bytecodes.c`

It looks like you have all the pieces, so that should be reasonably straightforward.

gvanrossum committed:
Format large families more nicely

gvanrossum reviewed (commented):
I'm not entirely sure where you'd like me to go with this. Clearly I've delivered more than you asked for. :-)

I will definitely remove the redundant cases from ceval.c before merging this, but I'd like to get the rest of the questions sorted first.

I could just keep bytecodes.c and write a much simpler script that transforms it into cases.h (basically just stripping the dummy stuff at the top and bottom and replacing `inst(...)` with `TARGET(...)`). But I'm not sure what would be the value of that?

Separately, do you want me to rename cases.h to generated_cases.c.h?

gvanrossum committed:
Rename cases.h to generated_cases.c.h

gvanrossum committed:
Merge remote-tracking branch 'origin/main' into generate-ceval-switch

gvanrossum committed:
Regenerate cases after merging latest main

gvanrossum committed:
Separate family members by commas

Also put 3 per line instead of 4.

gvanrossum committed:
Merge remote-tracking branch 'origin/main' into generate-ceval-switch

gvanrossum committed:
Regenerated files after merge from main

gvanrossum committed:
Get rid of the C parser infrastructure

Blocks are once more "parsed" by skipping matching brackets.
The algorithm for deciding when to add a DISPATCH() at the end
of a block is not always optimal, hopefully the extra DISPATCH()
calls generated will be recognized by the compilers as dead code.

gvanrossum committed:
Remove stack effect from bytecodes.c

gvanrossum committed:
Update README.md

gvanrossum committed:
Merge remote-tracking branch 'origin/main' into generate-ceval-switch

brandtbucher commented on Makefile.pre.in:
```diff
@@ -1445,7 +1445,19 @@ regen-opcode-targets:
 		$(srcdir)/Python/opcode_targets.h.new
```
I didn't review these changes.

brandtbucher commented on Python/generated_cases.c.h:
```diff
@@ -0,0 +1,3871 @@
+// This file is generated by Tools/scripts/generate_cases.py
```
This file should be added to the list of generated files in `.gitattributes`.

brandtbucher commented on Tools/cases_generator/README.md:
```diff
@@ -0,0 +1,33 @@
+# Tooling to generate interpreters
```
I didn't review any of the files in this directory for correctness... the output smells okay to me, I'm not a parsing expert, and I'm sure the code will change in the near future. :)

brandtbucher commented on Python/ceval.c:
```diff
@@ -1158,3868 +1158,7 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
+#include "generated_cases.c.h"
 
 #if USE_COMPUTED_GOTOS
         TARGET_DO_TRACING:
```
It would be nice if we could also wrap `DO_TRACING` and `_unknown_opcode`/`EXTRA_CASES` up in the same mechanism. A good project for later, probably.

brandtbucher commented on Python/bytecodes.c:
```diff
@@ -0,0 +1,3983 @@
+#define SET_TOP(v)        (stack_pointer[-1] = (v))
+#define GETLOCAL(i)     (frame->localsplus[i])
+
+#define inst(name, stack_effect) case name:
```
I got a nice big pile of red squiggles when I opened `bytecodes.c`. Here are all of the issues:

- The `inst` macro definition has a `stack effect` parameter that isn't used by any of the actual instructions. 
- Some labels are missing:
  - `exception_unwind`
  - `resume_frame`
  - `resume_with_error`
  - `start_frame`
- Some locals are missing:
  - `call_shape`
  - `first_instr`
  - `throwflag`
- A bunch of other stuff is missing:
  - `NAME_ERROR_MSG`
  - `PEEK`
  - `PyDictKeyEntry`
  - `PyDictOrValues`
  - `PyDictUnicodeEntry`
  - `PyModuleObject`
  - `_PyDictOrValues_GetValues`
  - `_PyListIterObject`
  - `_PyObject_DictOrValuesPointer`
  - `_PyOpcode_Deopt`
  - `_PyRangeIterObject`
  - `_Py_ID`
  - `_Py_STR`
  - `binary_ops`
  - `struct _dictkeysobject`
  - `struct _dictvalues`
  - `struct _is`

gvanrossum replied:
Good point. I added some more dummy arguments and labels, fixed the `inst()` macro, and I'm just copying all `#include`s from ceval.c...

brandtbucher commented on Python/bytecodes.c:
```diff
@@ -0,0 +1,3983 @@
+
+        // stack effect: ( -- )
+        inst(RESUME_QUICK) {
+            PREDICTED(RESUME_QUICK);
```
Autogenerating these `PREDICTED` targets (and *maybe* even the `PREDICT` part) would be a good future project too.

brandtbucher commented on Python/bytecodes.c:
```diff
@@ -0,0 +1,3983 @@
+#define GETLOCAL(i)     (frame->localsplus[i])
+
+#define inst(name, stack_effect) case name:
+#define family(name) static int family_##name
```
There's also lots of red in the `family` section. The `family` macro expands to `static int family_xxx = ...`. The C preprocessor expands all of the following opcode names into integers, so what C sees is `static int family_xxx = 123, 456, 789, ...`. It complains that `456` and `789` aren't identifiers, since the commas make it think you're trying to declare a bunch of variables. So either the macro or the DSL may need updating here.

gvanrossum replied:
Ah, that was a last-minute change, I had a `+` separator there first. I'll put it inside curlies and make the macro define an unsized array.

brandtbucher commented on Python/bytecodes.c:
```diff
@@ -0,0 +1,3983 @@
+family(binary_op) =
+    BINARY_OP, BINARY_OP_ADAPTIVE, BINARY_OP_ADD_FLOAT,
+    BINARY_OP_ADD_INT, BINARY_OP_ADD_UNICODE, BINARY_OP_INPLACE_ADD_UNICODE,
+    BINARY_OP_MULTIPLY_FLOAT, BINARY_OP_MULTIPLY_INT, BINARY_OP_SUBTRACT_FLOAT;
```
```suggestion
    BINARY_OP_MULTIPLY_FLOAT, BINARY_OP_MULTIPLY_INT, BINARY_OP_SUBTRACT_FLOAT,
```

brandtbucher commented on Python/bytecodes.c:
```diff
@@ -0,0 +1,3983 @@
+    CALL_BUILTIN_FAST_WITH_KEYWORDS, CALL_METHOD_DESCRIPTOR_FAST_WITH_KEYWORDS, CALL_NO_KW_BUILTIN_FAST,
+    CALL_NO_KW_BUILTIN_O, CALL_NO_KW_ISINSTANCE, CALL_NO_KW_LEN,
+    CALL_NO_KW_LIST_APPEND, CALL_NO_KW_METHOD_DESCRIPTOR_FAST, CALL_NO_KW_METHOD_DESCRIPTOR_NOARGS,
+    CALL_NO_KW_METHOD_DESCRIPTOR_O, CALL_NO_KW_STR_1, CALL_NO_KW_TUPLE_1;
```
```suggestion
    CALL_NO_KW_METHOD_DESCRIPTOR_O, CALL_NO_KW_STR_1, CALL_NO_KW_TUPLE_1,
```

brandtbucher commented on Python/bytecodes.c:
```diff
@@ -0,0 +1,3983 @@
+    COMPARE_OP_INT_JUMP, COMPARE_OP_STR_JUMP;
+family(extended_arg) = EXTENDED_ARG, EXTENDED_ARG_QUICK;
+family(for_iter) =
+    FOR_ITER, FOR_ITER_ADAPTIVE, FOR_ITER_LIST;
```
```suggestion
    FOR_ITER, FOR_ITER_ADAPTIVE, FOR_ITER_LIST,
```

brandtbucher commented on Python/bytecodes.c:
```diff
@@ -0,0 +1,3983 @@
+family(for_iter) =
+    FOR_ITER, FOR_ITER_ADAPTIVE, FOR_ITER_LIST;
+    FOR_ITER_RANGE;
+family(jump_backward) = JUMP_BACKWARD, JUMP_BACKWARD_QUICK;
```
This family will be removed once you resolve this morning's merge conflict (these opcodes were re-combined).

brandtbucher commented on Python/bytecodes.c:
```diff
@@ -0,0 +1,3983 @@
+family(load_global) =
+    LOAD_GLOBAL, LOAD_GLOBAL_ADAPTIVE, LOAD_GLOBAL_BUILTIN;
+    LOAD_GLOBAL_MODULE;
+family(resume) = RESUME, RESUME_QUICK;
```
This family will be removed once you resolve this morning's merge conflict (these opcodes were re-combined).

brandtbucher commented on Python/bytecodes.c:
```diff
@@ -0,0 +1,3983 @@
+    LOAD_ATTR, LOAD_ATTR_ADAPTIVE, LOAD_ATTR_CLASS,
+    LOAD_ATTR_GETATTRIBUTE_OVERRIDDEN, LOAD_ATTR_INSTANCE_VALUE, LOAD_ATTR_MODULE,
+    LOAD_ATTR_PROPERTY, LOAD_ATTR_SLOT, LOAD_ATTR_WITH_HINT,
+    LOAD_ATTR_METHOD_LAZY_DICT, LOAD_ATTR_METHOD_NO_DICT, LOAD_ATTR_METHOD_WITH_DICT;
```
```suggestion
    LOAD_ATTR_METHOD_LAZY_DICT, LOAD_ATTR_METHOD_NO_DICT, LOAD_ATTR_METHOD_WITH_DICT,
```

brandtbucher commented on Python/bytecodes.c:
```diff
@@ -0,0 +1,3983 @@
+family(load_const) = LOAD_CONST, LOAD_CONST__LOAD_FAST;
+family(load_fast) = LOAD_FAST, LOAD_FAST__LOAD_CONST, LOAD_FAST__LOAD_FAST;
+family(load_global) =
+    LOAD_GLOBAL, LOAD_GLOBAL_ADAPTIVE, LOAD_GLOBAL_BUILTIN;
```
```suggestion
    LOAD_GLOBAL, LOAD_GLOBAL_ADAPTIVE, LOAD_GLOBAL_BUILTIN,
```

brandtbucher commented on Python/bytecodes.c:
```diff
@@ -0,0 +1,3983 @@
+    STORE_ATTR_SLOT, STORE_ATTR_WITH_HINT;
+family(store_fast) = STORE_FAST, STORE_FAST__LOAD_FAST, STORE_FAST__STORE_FAST;
+family(store_subscr) =
+    STORE_SUBSCR, STORE_SUBSCR_ADAPTIVE, STORE_SUBSCR_DICT;
```
```suggestion
    STORE_SUBSCR, STORE_SUBSCR_ADAPTIVE, STORE_SUBSCR_DICT,
```

brandtbucher commented on Tools/cases_generator/parser.py:
```diff
@@ -0,0 +1,222 @@
+    end: int
+    owner: PLexer
+
+    def __rep__(self):
```
```suggestion
    def __repr__(self):
```

brandtbucher commented on Tools/cases_generator/parser.py:
```diff
@@ -0,0 +1,222 @@
+            begin = srclines.index("// BEGIN BYTECODES //")
+            end = srclines.index("// END BYTECODES //")
+            src = "\n".join(srclines[begin+1 : end])
+    else:
+        filename = None
+        src = "if (x) { x.foo; // comment\n}"
```
Not sure why this is here... maybe leftover debugging?

We'd probably want to just raise, right?

gvanrossum replied:
This entire `__main__` block is for debugging the parser, and this is some sample test input. :-)

brandtbucher commented on Tools/cases_generator/generate_cases.py:
```diff
@@ -0,0 +1,134 @@
+
+
+@dataclass
+class Instruction:
```
Minor: I'm not quite sure what this gives us over `parser.InstDef`?

gvanrossum replied:
Not much, I'll clean it up.

brandtbucher commented on Tools/cases_generator/parser.py:
```diff
@@ -0,0 +1,222 @@
+
+
+T = TypeVar("T", bound="Parser")
+S = TypeVar("S", bound="Node")
```
Minor: `P` and `N` would be clearer to me here.

brandtbucher commented on Tools/cases_generator/plexer.py:
```diff
@@ -0,0 +1,104 @@
+        else:
+            with open(filename) as f:
+                src = f.read()
+    else:
+        filename = None
+        src = "if (x) { x.foo; // comment\n}"
```
Leftover debugging?

brandtbucher reviewed (changes requested):
Thanks for tackling this! Most of this review is a bunch of things that made VSCode angry in `bytecodes.c`. I didn't spend *too* much time reviewing the parsing/lexing/generating code, but it seems like it's doing the right thing and is very much a work in progress. :)

+1 for merging after the issues in `bytecodes.c` are tidied up, or +0 for merging now and fixing those later.

brandtbucher commented on Python/bytecodes.c:
```diff
@@ -0,0 +1,3983 @@
+    }
+ handle_eval_breaker:;
+ unbound_local_error:;
+ error:;
```
These semicolons aren't needed, right?

```suggestion
 handle_eval_breaker:
 unbound_local_error:
 error:
```

gvanrossum replied:
Oh, you're right. It'll have to wait until another time. (The _last_ label does need the `;` though, a label must be followed by a statement.)

gvanrossum committed:
Add generated_cases.c.h to .gitattributes as generated file

gvanrossum committed:
Update from Brandt's code review

- Change family syntax to `family(NAME) = {OP1, OP2, ...};`
- Some cleanups in parser.py and generate_cases.py
- Fixes to bytecodes.c to reduce red wiggles in VS Code

gvanrossum committed:
Merge remote-tracking branch 'origin/main' into generate-ceval-switch

gvanrossum reviewed (commented):
I'm pushing the new version in a few minutes, after merging two tough commits from main. The pushed version will once again have the cases in ceval.c, I'll remove those later (right before merging).

gvanrossum committed:
Update generated files after merging from main

gvanrossum committed:
Remove the redundant cases from ceval.c

If you get a merge conflict, talk to Guido
